WooCommerce/Classes/ViewRelated/ReusableViews/UpdateProgressImage.swift

```diff
@@ -0,0 +1,74 @@
+import Foundation
+import SwiftUI
+
+extension UIImage {
+    static func composite(images: [UIImage]) -> UIImage? {
+        guard let firstImage = images.first else {
+            return nil
+        }
+        let size = firstImage.size
+        let rect = CGRect(origin: .zero, size: size)
+        UIGraphicsBeginImageContext(size)
+
+        for image in images {
+            image.draw(in: rect)
+        }
+
+        let result = UIGraphicsGetImageFromCurrentImageContext()
+        UIGraphicsEndImageContext()
+
+        return result
+    }
+
+    private static func softwareUpdateProgressFill(progress: CGFloat) -> UIImage? {
+        let rect = CGRect(x: 0, y: 0, width: Constants.size, height: Constants.size)
+        let clippingRect = CGRect(x: 0, y: (1 - progress) * Constants.size, width: Constants.size, height: Constants.size)
```

Comment:
Please consider constraining progress to avoid unexpected values of `y`

Reply:
Good call, added it in 1ef2c2f2d. I thought of adding a `precondition` there, but I'd rather not crash during a software update, so clamping feels like a better idea in this case. I still added an `assert`, which would crash on debug builds so we can notice any bugs with that